File: ml_update.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <time.h>
#include <sys/time.h>
#include <jubatus/client/recommender_client.hpp>
#include <jubatus/client/recommender_types.hpp>

using namespace std;
using namespace jubatus;
using namespace jubatus::client;
using namespace pfi::lang;

const string NAME = "recommender_ml";

int main(int argc, char* argv[]){
  
  string convert;
  ifstream ifc("config.json");
  stringstream ss;
  ss << ifc.rdbuf();

  config_data config;

  config.converter = ss.str();
  config.method = "lsh";

  recommender r("localhost", 9199, 1.0);
  r.set_config(NAME, config);

  ifstream ifs("./dat/ml-100k/u.data");
  if (!ifs){
    throw string ("cannot open data file");
  }

  string userid, movieid, rating, mtime;
  jubatus::datum d;
  int n = 0;
  while((ifs >> userid >> movieid >> rating >> mtime)!=0){
    d.num_values.clear();
    if (n % 1000 == 0)
       cout << n << endl;
    d.num_values.push_back(make_pair(movieid, pfi::lang::lexical_cast<int>(rating)));
    r.update_row(NAME, userid, d);
    n++;
  }
}

File: ml_update.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import sys, json
from jubatus.recommender import client
from jubatus.recommender import types

NAME = "recommender_ml";

if __name__ == '__main__':

    recommender = client.recommender("127.0.0.1",9199)

    converter = {
              "string_filter_types": {},
              "string_filter_rules": [],
                'num_filter_types': {},
                'num_filter_rules': [],
                'string_types': {},
                'string_rules': [],
                'num_types': {},
                'num_rules': [{"key" : "*", "type" : "num"}]
                }

    config = types.config_data("lsh", json.dumps(converter))
    recommender.set_config(NAME, config)

    n = 0
    for line in open('./dat/ml-100k/u.data'):
        userid, movieid, rating, mtime = line[:-1].split('\t')
        datum = types.datum(  [], [[str(movieid), float(rating)]] )
        if (n % 1000 == 0):
            print n
        recommender.update_row(NAME, userid, datum)
        n = n + 1;




